symforce/slam/imu_preintegration/manifold_symbolic.py

```diff
@@ -0,0 +1,346 @@
+            i.e., the vector which when subtracted from the accelerometer measurements gives the
+            true acceleration (up to bias and noise) of the IMU.
+        dt (T.Scalar): The time between timestep i and j: t_j - t_i
+        epsilon (T.Scalar): epsilon used for numerical stability
+    """
+    delta_gyro_bias = gyro_bias_i - gyro_bias_hat
+    delta_accel_bias = accel_bias_i - accel_bias_hat
+
+    # Correct preintegrated measurements for updated bias estimates
+    corrected_DR = DR * sf.Rot3.from_tangent(
+        (DR_D_gyro_bias * delta_gyro_bias).to_storage(), epsilon
+    )
+    corrected_Dv = Dv + Dv_D_gyro_bias * delta_gyro_bias + Dv_D_accel_bias * delta_accel_bias
+    corrected_Dp = Dp + Dp_D_gyro_bias * delta_gyro_bias + Dp_D_accel_bias * delta_accel_bias
+
+    res_R = (corrected_DR.inverse() * pose_i.R.inverse() * pose_j.R).to_tangent(epsilon)
+    res_v = pose_i.R.inverse() * (vel_j - vel_i - gravity * dt) - corrected_Dv
+    res_p = (
+        pose_i.R.inverse() * (pose_j.t - pose_i.t - vel_i * dt - sf.S(1) / 2 * gravity * dt ** 2)
+        - corrected_Dp
+    )
+
+    res = sf.V9(*res_R, *res_v, *res_p)
+
+    return sf.M91(sqrt_info.lower_triangle() * res)
+
+
+def right_jacobian(tangent: sf.V3, epsilon: T.Scalar) -> sf.M33:
+    """
+    The right jacobian J(v) is d/du Log(Exp(v)^{-1} Exp(v + u)), i.e., a matrix such that
+    Exp(v + dv) ~= Exp(v) Exp(J(v) dv), where v and u are tangent vectors of SO(3).
+
+    Returns J(tangent).
+    """
+    # NOTE(brad): We use sqrt(epsilon) to ensure cos(norm) is far enough from 1 to mitigate
+    # the effects of catastrophic cancellation in 1 - cos(norm) below.
+    norm = tangent.norm(sf.sqrt(epsilon))
+    tangent_hat = sf.Matrix.skew_symmetric(tangent)
+    out = sf.M33.eye()
+    out += ((1 - sf.cos(norm)) / (norm ** 2)) * tangent_hat
+    out += ((norm - sf.sin(norm)) / (norm ** 3)) * (tangent_hat * tangent_hat)
+    return out
+
+
+def handwritten_new_state_D_state_gyro_accel(
+    DR: sf.Rot3, corrected_gyro: sf.V3, corrected_accel: sf.V3, dt: T.Scalar, epsilon: T.Scalar
+) -> T.Tuple[sf.M99, sf.M93, sf.M93]:
+    """
+    Calculates the derivatives of the new state (meaning the DR, Dv, and Dp) w.r.t. the previous state,
+    gyroscope measurement, and the accelerometer mesaurement.
+
+
+    Args:
+        DR (sf.Rot3): Preintegrated DR of the previous state
+        corrected_gyro (sf.V3): gyroscope measurment corrected for IMU bias
+        corrected_accel (sf.V3): accelerometer measurement corrected for IMU bias
+        dt (T.Scalar): Time difference between the previous time step and the new time step
+        epsilon (T.Scalar): epsilon for numerical stability
+
+    Returns:
+        T.Tuple[sf.M99, sf.M93, sf.M93]: new_state_D_old_state, new_state_D_gyro_measurement, new_state_D_accel_measurement
+    """
+    # NOTE(Brad): I put each of the 3 derivatives in one function to make it easier to keep
+    # subexpressions similar to improve CSE
```

Comment:
This shouldn't matter since the update is generated as a single function no matter what, right?

Reply:
Hmm. Well, my thinking was that if you have common sub-expressions between the different derivatives (for example for matrices `A`, `B`, and `C`, `A * B * C` might appear in two different expressions), it would be worthwhile to take care to ensure that you do `(A * B) * C` in both, and not, say `(A * B) * C` in one and `A * (B * C)` (as even though they are the same, CSE won't recognize that).

But, now that I look at it again, I don't actually see any such examples in the derivatives as they are.

As such, I think the comment will just increase confusion for no real reason, so I'll remove it.